File: 20_Valid_Parentheses.py
class Solution(object):
    def isValid(self, s):
        paren_map = {')':'(', '}':'{', ']':'['}
        stack = [] 
        for i in s: 
            if i in "({[":
                stack.append(i) 
            if i in paren_map.keys():
                if len(stack) <= 0:
                    return False 
                lri = stack.pop()
                if paren_map[i] != lri:
                    return False 
        return len(stack) == 0 
    

File: 20_Valid_Parentheses.java
class Solution {
    
    static Map<Character, Character> mappings = new HashMap<>(); 
    
    static { 
            mappings.put('(', ')');
            mappings.put('{', '}');
            mappings.put('[', ']'); 
        }
    
    
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>(); 
        for(int i = 0; i < s.length(); i++)
        {
            char c = s.charAt(i); 
            if(mappings.containsKey(c))
            {
                stack.push(mappings.get(c)); 
            }
            else if (mappings.containsValue(c))
            {
                if(stack.isEmpty() || stack.pop() != c) return false;
            }
        }
        return stack.isEmpty(); 
    }
}
